File: 1814-count-nice-pairs-in-an-array/1814-count-nice-pairs-in-an-array.py
class Solution:
    def countNicePairs(self, nums: List[int]) -> int:
        def rev(x: int) -> int:
            stack = []
            while x > 0:
                stack.append(x % 10)
                x //= 10
            num, digit = 0, 0
            while stack:
                num = (num + stack.pop() * (10 ** digit)) % (10 ** 9 + 7)
                digit += 1
            return num

        ans, pairs = 0, dict()
        for num in nums:
            diff = num - rev(num)
            pairs[diff] = pairs.get(diff, 0) + 1
        
        for _, freq in pairs.items():
            ans = (ans + ((freq * (freq - 1)) >> 1)) % (10 ** 9 + 7)
        
        return ans

File: 1814-count-nice-pairs-in-an-array/1814-count-nice-pairs-in-an-array.cpp
class Solution {
private:
    int MOD = 1'000'000'007;
    int rev(int num){
        int res = 0;
        while (num > 0){
            res = res * 10 + num % 10;
            num /= 10;
        }
        
        return res;
    }
    
public:
    int countNicePairs(vector<int>& nums) {
        vector<int> arr;
        for (int num : nums)
            arr.push_back(num - rev(num));
        
        int ans = 0;
        unordered_map<int, int> pairs;
        for (int num : arr){
            ans = (ans + pairs[num]) % MOD;
            ++pairs[num];
        }
        
        return ans;
    }
};